crates/rattler/src/libsolv/pool.rs

```diff
@@ -0,0 +1,300 @@
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Create repo from a pool
+    pub fn create_repo<S: AsRef<str>>(&mut self, url: S) -> Repo {
+        unsafe {
+            let c_url = c_string(url);
+            Repo::new(
+                NonNull::new(ffi::repo_create(self.as_ptr().as_mut(), c_url.as_ptr()))
+                    .expect("libsolv repo_create returned nullptr"),
+            )
+        }
+    }
+
+    /// Create the solver
+    pub fn create_solver(&mut self) -> Solver {
+        let solver = NonNull::new(unsafe { ffi::solver_create(self.as_ptr().as_mut()) })
+            .expect("solver_create returned a nullptr");
+        Solver::new(solver)
+    }
+
+    /// Create the whatprovides on the pool which is needed for solving
+    pub fn create_whatprovides(&mut self) {
+        // Safe because pointer must exist
+        unsafe {
+            ffi::pool_createwhatprovides(self.as_ptr().as_mut());
+        }
+    }
+}
+
+/// Intern string like types
+fn intern_str<T: AsRef<str>>(pool: &mut PoolRef, str: T) -> StringId {
+    // Safe because conversion is valid
+    let c_str = CString::new(str.as_ref()).expect("could never be null because of trait-bound");
+    let length = c_str.as_bytes().len();
+    let c_str = c_str.as_c_str();
+
+    // Safe because pool exists and function accepts any string
+    unsafe {
+        StringId(ffi::pool_strn2id(
+            pool.as_ptr().as_mut(),
+            c_str.as_ptr(),
+            length.try_into().expect("string too large"),
+            1,
+        ))
+    }
+}
+
+/// Finds a previously interned string or returns `None` if it wasnt found.
+fn find_intern_str<T: AsRef<str>>(pool: &PoolRef, str: T) -> Option<StringId> {
+    // Safe because conversion is valid
+    let c_str = CString::new(str.as_ref()).expect("could never be null because of trait-bound");
+    let length = c_str.as_bytes().len();
+    let c_str = c_str.as_c_str();
+
+    // Safe because pool exists and function accepts any string
+    unsafe {
+        let id = ffi::pool_strn2id(
+            pool.as_ptr().as_ptr(),
+            c_str.as_ptr(),
+            length.try_into().expect("string too large"),
+            0,
+        );
+        if id == 0 {
+            None
+        } else {
+            Some(StringId(id))
+        }
+    }
+}
+
+/// Interns from Target type to Id
+pub trait Intern {
+    type Id;
+
+    /// Interns the type in the [`Pool`]
+    fn intern(&self, pool: &mut PoolRef) -> Self::Id;
+}
+
+pub trait FindInterned: Intern {
+    /// Finds a previously interned instance in the specified [`Pool`]
+    fn find_interned_id(&self, pool: &PoolRef) -> Option<Self::Id>;
+}
+
+/// Wrapper for the StringId of libsolv
+#[derive(Copy, Clone)]
+pub struct StringId(pub(super) ffi::Id);
+
+impl StringId {
+    /// Resolve to the interned type returns a string reference
+    pub fn resolve<'a>(&self, pool: &'a PoolRef) -> &'a str {
```

Comment:
This could fail if you create two different pools and use the Id of Pool a to retrieve a string from Pool b. 

I think the c_str, in that case, will be `nullptr` so I think this method should return an Option.

Reply:
Yeah, so it seems to be kind of strange behavior by libsolv. Maybe you can have a look :)

Reply:
Fixed